gpcontrib/arenadata_toolkit/src/track_files.c

```diff
@@ -0,0 +1,1228 @@
+		char		relstorage;
+		HeapTuple	pg_class_tuple;
+		uint64_t	hash;
+
+		if (!state->scan)
+			break;
+
+		pg_class_tuple = systable_getnext(state->scan);
+
+		if (!HeapTupleIsValid(pg_class_tuple))
+		{
+			systable_endscan(state->scan);
+			heap_close(state->pg_class_rel, AccessShareLock);
+			state->scan = NULL;
+			state->pg_class_rel = NULL;
+			break;
+		}
+
+		datums[6] = heap_getattr(pg_class_tuple, Anum_pg_class_relnamespace, RelationGetDescr(state->pg_class_rel), &nulls[6]);
+		relnamespace = DatumGetObjectId(datums[6]);
+
+		if (!schema_is_tracked(relnamespace))
+			continue;
+
+		datums[7] = heap_getattr(pg_class_tuple, Anum_pg_class_relkind, RelationGetDescr(state->pg_class_rel), &nulls[7]);
+		relkind = CharGetDatum(datums[7]);
+
+		if (!relkind_is_tracked(relkind))
+			continue;
+
+		datums[8] = heap_getattr(pg_class_tuple, Anum_pg_class_relstorage, RelationGetDescr(state->pg_class_rel), &nulls[8]);
+		relstorage = CharGetDatum(datums[8]);
+
+		if (!relstorage_is_tracked(relstorage))
+			continue;
+
+		datums[0] = ObjectIdGetDatum(HeapTupleGetOid(pg_class_tuple));
+
+		datums[1] = heap_getattr(pg_class_tuple, Anum_pg_class_relname, RelationGetDescr(state->pg_class_rel), &nulls[1]);
+
+		datums[2] = heap_getattr(pg_class_tuple, Anum_pg_class_relfilenode, RelationGetDescr(state->pg_class_rel), &nulls[2]);
+		filenode = DatumGetObjectId(datums[2]);
+
+		if (nulls[2])
+			continue;
+
+		/* Bloom filter check */
+		hash = bloom_set_calc_hash(&filenode, sizeof(filenode));
+		if (!bloom_isset(tf_get_global_state.bloom, hash))
+			continue;
+
+		/*
+		 * Taking a lock and calling relation_open in dbsize_calc_size is
+		 * quite suboptimal. The size calculation strategy should be revised
+		 * in future.
+		 */
+		datums[3] = Int64GetDatum(dbsize_calc_size(HeapTupleGetOid(pg_class_tuple)));
+		datums[4] = CharGetDatum(tf_get_global_state.bloom->is_set_all ? 'i' : 'a');
+		datums[5] = Int32GetDatum(GpIdentity.segindex);
+
+		result = heap_form_tuple(funcctx->tuple_desc, datums, nulls);
+
+		SRF_RETURN_NEXT(funcctx, HeapTupleGetDatum(result));
+	}
+
+	while (true)
+	{
+		Oid			filenode;
+
+		if (!tf_get_global_state.next_drop)
+			break;
+
+		filenode = lfirst_oid(tf_get_global_state.next_drop);
+		tf_get_global_state.next_drop = lnext(tf_get_global_state.next_drop);
+
+		nulls[0] = true;
+		nulls[1] = true;
+		datums[2] = filenode;
+		datums[3] = Int64GetDatum(0);
+		datums[4] = CharGetDatum('d');
+		datums[5] = Int32GetDatum(GpIdentity.segindex);
+		nulls[6] = true;
+		nulls[7] = true;
+		nulls[8] = true;
+
+		result = heap_form_tuple(funcctx->tuple_desc, datums, nulls);
+
+		SRF_RETURN_NEXT(funcctx, HeapTupleGetDatum(result));
+	}
+
+	SRF_RETURN_DONE(funcctx);
+}
+
+/*
+ * Function used in "arenadata_toolkit.tables_track" view. In order to keep bloom filter
+ * in consistent state across segments this function dispatches main tracking logic to the
+ * segments in a distributed transaction.
+ */
+Datum
+tracking_get_track(PG_FUNCTION_ARGS)
```

Comment:
If one tries to call extension functions before the shared module is inited, he may get a crash (tested with this function).
Better to guard functions of the extension and return a proper error in such case, for ex. smth like the sample code below does:
```diff
diff --git a/gpcontrib/arenadata_toolkit/src/tf_shmem.c b/gpcontrib/arenadata_toolkit/src/tf_shmem.c
index 9615c60b86..b18cd53ada 100644
--- a/gpcontrib/arenadata_toolkit/src/tf_shmem.c
+++ b/gpcontrib/arenadata_toolkit/src/tf_shmem.c
@@ -7,7 +7,7 @@
 #include "bloom_set.h"
 
 static shmem_startup_hook_type next_shmem_startup_hook = NULL;
-tf_shared_state_t *tf_shared_state;
+tf_shared_state_t *tf_shared_state = NULL;
 
 static Size
 tf_shmem_calc_size(void)
diff --git a/gpcontrib/arenadata_toolkit/src/track_files.c b/gpcontrib/arenadata_toolkit/src/track_files.c
index 3c34e4e7c5..c15c404317 100644
--- a/gpcontrib/arenadata_toolkit/src/track_files.c
+++ b/gpcontrib/arenadata_toolkit/src/track_files.c
@@ -520,6 +520,8 @@ tracking_get_track_main(PG_FUNCTION_ARGS)
        SRF_RETURN_DONE(funcctx);
 }
 
+
+extern tf_shared_state_t *tf_shared_state;
 /*
  * Function used in "arenadata_toolkit.tables_track" view. In order to keep bloom filter
  * in consistent state across segments this function dispatches main tracking logic to the
@@ -534,6 +536,11 @@ tracking_get_track(PG_FUNCTION_ARGS)
        Datum           values[9];
        bool            nulls[9] = {0};
 
+       if (tf_shared_state == NULL)
+       {
+               elog(ERROR, "module was not initialized properly");
+       }
+
        if (SRF_IS_FIRSTCALL())
        {
                MemoryContext oldcontext = CurrentMemoryContext;
```